### PR TITLE
External cmake project integration

### PR DESCRIPTION
**Pull request by bpkroth**
Work to make it easier/possible to integrate the C++ components of MLOS with an external C++ project (e.g. LevelDB, MySql, etc.) via `cmake`.

- Fixes some issues with our own `cmake` config assumptions.
- Adds examples patterns and documentation for using in an "external project" (currently just a minimal example).
- Hooks those into the CI pipelines.

This branch was split off of the work from #177 (since merged) to build NuGet packages for some of the C# components, so many of the original commits are shared.

- [x] Fix the hang in the simple example project when using Mlos without an agent.
  - This was the result of left over shared memory /dev/shm files from a previous run.  Should be fixed by inclusion of anonymous shared memory (#183).  For now, I've made the `Makefile`'s `test` target call the cleanup helper scripts.
- [x] Hook it into CI pipelines.
- [x] Lint check the markdown.
- Hook it into LevelDb (next PR)
  - `git clone` a particular version
  - apply a patch stored in our repo to do the integration
  - build/test

### CODE REVIEW

File: external/ExternalIntegrationExample/ExternalIntegrationExample.cpp [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    // Terminate the feedback channel.
+    //
+    mlosContext.TerminateFeedbackChannel();
+
+    // At this point there are no active feedback channel reader threads.
+    // Now, terminate the control channel.
+    //
+    mlosContext.TerminateControlChannel();
+
+    // FIXME: It's hanging ...
```
Comment (bpkroth) [inline external/ExternalIntegrationExample/ExternalIntegrationExample.cpp:121]: TODO Fix this ...

File: external/ExternalIntegrationExample/cmake/Common.cmake [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# This file would contain whatever common build definitions the external project needed.
+
+# TODO: TESTME:
```
Comment (bpkroth) [inline external/ExternalIntegrationExample/cmake/Common.cmake:3]: TODO: test this ...

File: external/ExternalIntegrationExample/.vscode/launch.json [hunk 3]
```diff
@@ -0,0 +1,23 @@
+{
+    // Use IntelliSense to learn about possible attributes.
+    // Hover to view descriptions of existing attributes.
+    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
+    "version": "0.2.0",
+    "configurations": [
+        {
+            "name": "(gdb) Attach - ExternalIntegrationExample",
+            "type": "cppdbg",
```
Comment (grlap) [inline external/ExternalIntegrationExample/.vscode/launch.json:9]: lldb ??

File: external/ExternalIntegrationExample/ExternalIntegrationExample.cpp [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+    // Create a feedback channel receiver thread.
+    //
+    ISharedChannel& feedbackChannel = mlosContext.FeedbackChannel();
+
+    // This background thread uses a lambda to monitor the feedback channel for
+    // new messages and process them using the callbacks registered for each
+    // message type in the global dispatch table.
+    //
+    std::future<bool> feedbackChannelReader = std::async(
```
Comment (grlap) [inline external/ExternalIntegrationExample/ExternalIntegrationExample.cpp:51]: we should switch to std::thread

File: external/ExternalIntegrationExample/.vscode/launch.json (hunk 3 above)
Comment (bpkroth) [reply external/ExternalIntegrationExample/.vscode/launch.json:9]: I can try that.  I just took whatever vscode popped up as the default.

File: external/ExternalIntegrationExample/ExternalIntegrationExample.cpp (hunk 4 above)
Comment (bpkroth) [reply external/ExternalIntegrationExample/ExternalIntegrationExample.cpp:51]: This file is basically just a stripped down copy of `SmartCache/Main.cpp`.  I could change this one or we could leave that for a future PR.
I wasn't really aiming to test the C++ code itself - I feel like we have better more well documented examples of that already.  This was just to get us started with making sure we could reference all of the MLOS code enought to build/test it with a non-MLOS project.

File: build/CMakeHelpers/ParseCsProjForCsFiles.sh [hunk 5]
```diff
@@ -7,7 +7,7 @@
 # CMakeLists.txt wrappers around .csproj files.
 
 cat *.csproj 2>/dev/null \
-    | egrep '<(Compile|SettingsRegistryDef) Include=".*\.cs"' \
+    | egrep '<(Compile|SettingsRegistryDef|MlosSettingsRegistryDefinition) Include=".*\.cs"' \
```
Comment (bpkroth) [inline build/CMakeHelpers/ParseCsProjForCsFiles.sh:10]: We renamed the `Item` name used in`SettingsRegistry.csproj` files for external projects.  This allows us to reuse that script for both when using `FetchContent` `cmake` integration.

File: build/CMakeHelpers/ParseCsProjForCsProjs.sh [hunk 6]
```diff
@@ -6,7 +6,7 @@
 # We use this to help dynamically build up the set of dependencies for our
 # CMakeLists.txt wrappers around .csproj files.
 
-cat *.csproj 2>/dev/null \
+cat *.csproj *.proj 2>/dev/null \
```
Comment (bpkroth) [inline build/CMakeHelpers/ParseCsProjForCsProjs.sh:9]: Let's us add `ProjectReference`s as dependencies for `.proj` files as well (see `Mlos.NetCore.Components.Packages/CMakeLists.txt`)

File: build/Mlos.NetCore.cmake [hunk 7]
```diff
@@ -114,11 +114,10 @@ function(add_mlos_dotnet_project)
         # we compose the dependency graph already above, so we can skip
         # building project references here in order to avoid some parallel
         # dotnet processes accessing the same files.
-        #
+        COMMAND ${DOTNET} build -m --configuration ${CMAKE_BUILD_TYPE} /p:BuildProjectReferences=false "${CSPROJ}"
         # Also, "dotnet build" doesn't update timestamps in a make compatible
         # way, so we also mark the projects as having been built using touch.
-        COMMAND ${DOTNET} build -m --configuration ${CMAKE_BUILD_TYPE} /p:BuildProjectReferences=false "${CSPROJ}" &&
-            touch -c "${CSPROJ}" "${OUTPUT_DLL}" "${BINPLACE_DLL}"
+        COMMAND ${CMAKE_COMMAND} -E touch_nocreate "${CSPROJ}" "${OUTPUT_DLL}" "${BINPLACE_DLL}"
```
Comment (bpkroth) [inline build/Mlos.NetCore.cmake:120]: Switch to using the more portable version of `touch` via `cmake`'s built in commands.

File: Dockerfile [hunk 8]
```diff
@@ -113,6 +115,8 @@ VOLUME /src/MLOS/target
 VOLUME /src/MLOS/tools
 VOLUME /src/MLOS/temp
 
+VOLUME /src/MLOS/external/ExternalIntegrationExample/build
```
Comment (bpkroth) [inline Dockerfile:118]: These aren't strictly necessary, but they make testing on multiple docker instances at the same time from a single source tree possible.

**Review by grlap — APPROVED** [review]

File: source/Mlos.NetCore.Components.Packages/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    OUTPUT_VARIABLE CS_PROJS
+    #COMMAND_ECHO STDERR
+)
+
+set(PROJ "${PROJECT_SOURCE_DIR}/${PROJECT_NAME}.proj")
+set(PKG_OUTPUT_DIR "${MLOS_ROOT}/target/pkg/${CMAKE_BUILD_TYPE}")
+
+set(DEPENDENCIES
+    ${PROJ}
+    #${CS_PROJS}
```
Comment (bpkroth) [inline source/Mlos.NetCore.Components.Packages/CMakeLists.txt:28]: FIXME: this is causing parallel build issues

**Pull request merged**